cpp/src/arrow/compute/kernels/row_encoder_internal.h

```diff
@@ -424,49 +525,189 @@ struct ListFixedWidthChildEncoder : KeyEncoder {
-        encoded_bytes[i] += child_width;
-        writer.Next();
       }
-      current_child_offset += list_length;
+      return encoded_ptr - encoded_ptr_start;
+    };
+
+    EncodeList<T>(data, batch_length, encoded_bytes,
+      getListBufferSizeWrapper, encodeList);
+    return Status::OK();
+  }
+
+  void EncodeNull(uint8_t** encoded_bytes) override {
+    EncodeVarLengthHeader<Offset>(*encoded_bytes, kNullByte);
+  }
+
+  Result<std::shared_ptr<ArrayData>> Decode(uint8_t** encoded_bytes, int32_t length,
+                                            MemoryPool* pool) override {
+    std::shared_ptr<Buffer> null_buf;
+    int32_t null_count;
+    ARROW_RETURN_NOT_OK(DecodeNulls(pool, length, encoded_bytes, &null_buf, &null_count));
+
+    // computing total lengths to allocate buffers
+    Offset list_length_sum = 0;
+    ChildOffset child_length_sum = 0;
+    for (int64_t i = 0; i < length; i++) {
+      auto encoded_ptr = encoded_bytes[i];
+      Offset list_length = util::SafeLoadAs<Offset>(encoded_bytes[i]);
+      encoded_ptr += sizeof(Offset);
+      for (int64_t j = 0; j < list_length; j++) {
+        ChildOffset child_length =
+          util::SafeLoadAs<ChildOffset>(encoded_ptr + kExtraByteForNull);
+        child_length_sum += child_length;
+        encoded_ptr += kExtraByteForNull + sizeof(ChildOffset) + child_length;
+      }
+      list_length_sum += list_length;
     }
-    raw_offsets[length] = current_child_offset;
-    writer.Finish();
+
+    ARROW_ASSIGN_OR_RAISE(auto offset_buf,
+                          AllocateBuffer(sizeof(Offset) * (1 + length), pool));
+    ARROW_ASSIGN_OR_RAISE(auto child_key_buf, AllocateBuffer(child_length_sum, pool));
+    ARROW_ASSIGN_OR_RAISE(auto child_validity_buf,
+                          AllocateBitmap(child_length_sum, pool));
+    ARROW_ASSIGN_OR_RAISE(auto child_offset_buf,
+                          AllocateBuffer(sizeof(ChildOffset) * (1 + list_length_sum)));
+
+    auto raw_offset = reinterpret_cast<Offset*>(offset_buf->mutable_data());
+    auto raw_child_key = child_key_buf->mutable_data();
+    auto raw_child_validity = child_validity_buf->mutable_data();
+    auto raw_child_offset =
+      reinterpret_cast<ChildOffset*>(child_offset_buf->mutable_data());
+
+    int64_t child_null_count = 0;
+    ChildOffset child_length_cumsum = 0;
+    ListDecoder<T>(encoded_bytes, length, child_length_sum,
+      child_null_count, raw_offset, raw_child_validity,
+    [&](int64_t i, int64_t){
+      ChildOffset child_length = util::SafeLoadAs<ChildOffset>(encoded_bytes[i]);
+      encoded_bytes[i] += sizeof(ChildOffset);
+
+      memcpy(raw_child_key, encoded_bytes[i], child_length);
+      raw_child_key += child_length;
+      encoded_bytes[i] += child_length;
+
+      util::SafeStore<ChildOffset>(raw_child_offset, child_length_cumsum);
+      child_length_cumsum += child_length;
+      raw_child_offset++;
+    }, [&](int64_t i, int64_t){
+      encoded_bytes[i] += sizeof(ChildOffset);
```

Comment:
should we also memset `raw_child_key` to 0?

Reply:
iirc if a varlength is null, then we can set the offsets to be the range [x, x), meaning that it takes up no space in the `raw_child_key` array, hence we don't set it to 0.